expo/eas-cli
# [24/n][eas-cli] Move graphql and session into context

# Checklist


# Why

There is one more major piece that can be moved to context: the graphql client. By moving this to the context:
- Global shared state is reduced (or eliminated). Previously the current user and graphql client were stored globally, and whatever was in the current session (access token or session token) was injected into the graphql request on every call no matter whether the call needed to be authenticated or not.
- We can more accurately determine which commands need to be logged-in. Previously this was done with a member variable and by moving this into the context we can determine this with 100% accuracy (since the tokens that signify being logged-in are only used in graphql).
- We can further isolate the internals of the authentication mechanisms from commands that don't need to access them. This provides the more complex authentication API only to commands that need it (login, logout) and the rest of the other commands now don't need access to anything but the graphql client from the context.

# How

This PR actually does more than just move the graphql client to a context member, but it does these changes procedurally by solely using the typechecker.

Here's how this PR was generated:
1. Remove `graphqlClient` from the global scope. Inject it in via method arguments all the way up and down the call tree in places that require it.
2. Make `LoggedInContextField` (replacement for `ActorContextField`) that requires login and exposes the actor and a logged-in graphql client (a client configured to use the access token / session secret belonging the actor). 
3. Use `yarn tsc` to make all replacements and pipe the graphql field all the way from the context down to the leaves (usually the queries or mutations). 
4. There are a couple special cases that couldn't be handled by a direct replacement and they are:
    1. Some Graphql requests take place during authentication and have custom auth header code (namely what used to be in `loginAsync`). These had to create their own graphql clients without authentication, which is more correct since we don't want to share urql caches between non-authenticated and authenticated requests.
    2. A couple more cases (grep for `createGraphqlClient`) that need special graphql clients since they are used in other context cases. (will look into ways to share objects across contexts in the future)
5. After this, it is clear that all graphql clients are created in places that don't require global state about the user session, so the user session (and all the session orchestration code) is moved into a new `SessionManager` object. This object is created by EasCommand and injected down to context objects that need it. This does a couple things: reduces global state (making testing easier), and makes it clearer where the session is accessible and what actions can be done (change methods to private).
6. This also makes the ApiV2 client an object that takes a session configuration (much like graphql). This client actually is never configured to make authenticated requests though, since all APIs it communicates with are logged-out APIs.

The end state is as follows:
1. EasCommand creates a single `SessionManager` object to be used by the running application. It is injected into command contexts.
2. `login` and `logout` commands directly request the session manager context to mutate it. All other commands don't need it.
3. `LoggedInContext` (and other contexts) use the injected session manager to transform it into things like authenticated graphql clients.
4. Other commands use `LoggedInContext` to get their graphql client. They then use it directly or pass it into the functions that they call that use it.

This allows for a few things:
- Better testability. Since these are now injected, we can mock them (using ts-mockito) without needing to mock them globally (using jest mock).
- Better authentication guarantees. Because authentication is no longer implicitly injected into the API calls, we can have more control over which ones require authentication and ensure they are authenticated. This prevents things like accidentally making authenticated API calls in command teardowns where the command may have logged-out for example. It also allows us to have multiple active stored user sessions in the future if we ever need it.

# Test Plan

- Run all changed commands, ensure their auth and graphql features are working.
- Run all tests.
